File: apps/web/src/api/coupleRhythm.ts
import { apiClient } from './client'

/**
 * Der wöchentliche Check-in – das kleine feste Ritual des Paarraums.
 *
 * Drei Fragen, eine Antwort je Person und Woche. `visible: false` heißt: ausgefüllt, aber
 * noch verdeckt, weil du selbst noch nicht dran warst. Das ist Absicht, keine Panne.
 */
export interface CoupleCheckinEntry {
  user_id: string
  name: string
  is_own: boolean
  done: boolean
  mood: string | null
  highlight: string | null
  wish: string | null
  visible: boolean
}

export interface CoupleCheckinWeek {
  week_start: string
  entries: CoupleCheckinEntry[]
  own_done: boolean
  both_done: boolean
  /** Der Wortlaut der drei Fragen – kommt vom Server, damit beide Seiten dasselbe sagen. */
  questions: Record<string, string>
  moods: Record<string, string>
}

export interface CoupleCheckinHistoryWeek {
  week_start: string
  moods: { user_id: string; name: string; mood: string | null; is_own: boolean }[]
}

export const coupleRhythmApi = {
  getCheckin: (coupleId: string) =>
    apiClient.get<CoupleCheckinWeek>(`/couple/links/${coupleId}/checkin`).then(r => r.data),

  saveCheckin: (
    coupleId: string,
    body: { mood?: string | null; highlight?: string | null; wish?: string | null },
  ) => apiClient.put<CoupleCheckinWeek>(`/couple/links/${coupleId}/checkin`, body).then(r => r.data),

  history: (coupleId: string) =>
    apiClient.get<CoupleCheckinHistoryWeek[]>(`/couple/links/${coupleId}/checkin/history`)
      .then(r => r.data),
}

/**
 * Wertschätzung – der zweite wiederkehrende Anlass.
 *
 * Bewusst ohne die Blindheitsregel des Check-ins: Sie geht sofort hinüber, auch wenn
 * nichts zurückkommt. Ein Geschenk, kein Zug im Wechselspiel.
 */
export interface CoupleAppreciation {
  id: string
  from_user_id: string
  from_name: string
  is_own: boolean
  body: string
  created_at: string
  seen_at: string | null
}

export interface CoupleAppreciationWall {
  received: CoupleAppreciation[]
  given: CoupleAppreciation[]
  unseen: number
  /** Anstöße für den leeren Zettel. */
  prompts: string[]
  partner_name: string | null
  max_chars: number
}

export const coupleAppreciationApi = {
  wall: (coupleId: string) =>
    apiClient.get<CoupleAppreciationWall>(`/couple/links/${coupleId}/wertschaetzung`)
      .then(r => r.data),

  leave: (coupleId: string, body: string) =>
    apiClient.post<CoupleAppreciation>(`/couple/links/${coupleId}/wertschaetzung`, { body })
      .then(r => r.data),

  markSeen: (coupleId: string) =>
    apiClient.post<{ seen: number }>(`/couple/links/${coupleId}/wertschaetzung/gesehen`)
      .then(r => r.data),
}

/**
 * Stimmungsbarometer – ein Regler von 1 bis 10, den die andere Person immer sieht.
 *
 * Bewusst ohne Blindheitsregel: Ein Barometer, das man erst nach eigener Eingabe sieht,
 * wäre keines. Der Verlauf ist dagegen nur der eigene – der aktuelle Stand ist ein Signal
 * zum Hinsehen, eine Chronik der schlechten Tage wäre Material für Vorhaltungen.
 */
export interface CoupleBarometerEntry {
  user_id: string | null
  name: string
  is_own: boolean
  /** null = diese Person hat den Regler noch nie gestellt. */
  value: number | null
  label: string | null
  note: string | null
  updated_at: string | null
}

export interface CoupleBarometerState {
  entries: CoupleBarometerEntry[]
  own_history: { value: number; created_at: string }[]
  /** Zahl → Wort. Eine Zahl ohne Wort wäre eine Note. */
  levels: Record<string, string>
  note_max_chars: number
}

export const coupleBarometerApi = {
  get: (coupleId: string) =>
    apiClient.get<CoupleBarometerState>(`/couple/links/${coupleId}/barometer`).then(r => r.data),

  set: (coupleId: string, value: number, note?: string | null) =>
    apiClient.put<CoupleBarometerState>(`/couple/links/${coupleId}/barometer`, { value, note })
      .then(r => r.data),
}

File: services/api/app/api/v1/routers/couple_rhythm.py
"""Router: die wiederkehrenden Anlaesse im Paarraum (Paartherapie, Rhythmus).

  GET  /couple/links/{couple_id}/checkin           - Check-in der laufenden Woche
  PUT  /couple/links/{couple_id}/checkin           - eigene Antwort speichern
  GET  /couple/links/{couple_id}/checkin/history   - die letzten Wochen als Stimmungs-Zeitstrahl
  GET  /couple/links/{couple_id}/wertschaetzung    - beide Richtungen
  POST /couple/links/{couple_id}/wertschaetzung    - einen Satz dalassen
  POST /couple/links/{couple_id}/wertschaetzung/gesehen - den "neu"-Hinweis abhaken
  GET  /couple/links/{couple_id}/barometer         - beide Regler + eigener Verlauf
  PUT  /couple/links/{couple_id}/barometer         - eigenen Regler stellen

Sicherheit: alles ueber ``require_couple_member`` im Service (404 fuer Fremde). Kein Zugriff
auf Fall-Daten, kein Echo-Aufruf - der Check-in ist bewusst ein reines Zwischen-euch-Ritual.
"""
from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends

from app.core.dependencies import get_current_user, get_pool
from app.schemas.couple_checkin import (
    CoupleAppreciation,
    CoupleAppreciationCreate,
    CoupleAppreciationWall,
    CoupleBarometerSet,
    CoupleBarometerState,
    CoupleCheckinHistoryWeek,
    CoupleCheckinSave,
    CoupleCheckinWeek,
)
from app.services import couple_appreciation_service as cas_appr
from app.services import couple_barometer_service as cbar
from app.services import couple_checkin_service as ccs
from app.services import couple_notify_service as notify
from app.services import couple_progress_service as progress

router = APIRouter(prefix="/couple/links/{couple_id}", tags=["couple-rhythm"])


@router.get("/checkin", response_model=CoupleCheckinWeek)
async def get_checkin(
    couple_id: UUID, current=Depends(get_current_user), pool=Depends(get_pool),
) -> CoupleCheckinWeek:
    async with pool.acquire() as conn:
        return CoupleCheckinWeek(**await ccs.load_week(conn, couple_id, current["user_id"]))


@router.put("/checkin", response_model=CoupleCheckinWeek)
async def save_checkin(
    couple_id: UUID, body: CoupleCheckinSave,
    current=Depends(get_current_user), pool=Depends(get_pool),
) -> CoupleCheckinWeek:
    user_id = current["user_id"]
    async with pool.acquire() as conn:
        await ccs.save(
            conn, couple_id, user_id,
            moods=body.moods, highlight=body.highlight, wish=body.wish,
        )
        woche = ccs.week_start()
        await progress.award(conn, couple_id, user_id, "checkin_done", woche.isoformat())
        state = await ccs.load_week(conn, couple_id, user_id)
        if not state["both_done"]:
            await notify.to_partner(conn, couple_id, user_id, notify.checkin_done())
        return CoupleCheckinWeek(**state)


@router.get("/checkin/history", response_model=list[CoupleCheckinHistoryWeek])
async def checkin_history(
    couple_id: UUID, current=Depends(get_current_user), pool=Depends(get_pool),
) -> list[CoupleCheckinHistoryWeek]:
    async with pool.acquire() as conn:
        rows = await ccs.load_history(conn, couple_id, current["user_id"])
    return [CoupleCheckinHistoryWeek(**r) for r in rows]


# ── Wertschaetzung ──────────────────────────────────────────────────────────
# Bewusst ohne die Blindheitsregel des Check-ins: ein Geschenk, kein Zug im Wechselspiel.


@router.get("/wertschaetzung", response_model=CoupleAppreciationWall)
async def get_appreciations(
    couple_id: UUID, current=Depends(get_current_user), pool=Depends(get_pool),
) -> CoupleAppreciationWall:
    async with pool.acquire() as conn:
        data = await cas_appr.load_wall(conn, couple_id, current["user_id"])
    return CoupleAppreciationWall(**data)


@router.post("/wertschaetzung", response_model=CoupleAppreciation, status_code=201)
async def leave_appreciation(
    couple_id: UUID, body: CoupleAppreciationCreate,
    current=Depends(get_current_user), pool=Depends(get_pool),
) -> CoupleAppreciation:
    user_id = current["user_id"]
    async with pool.acquire() as conn:
        row = await cas_appr.leave(conn, couple_id, user_id, body.body)
        await progress.award(conn, couple_id, user_id, "appreciation_left", row["id"])
        await notify.to_partner(conn, couple_id, user_id, notify.appreciation_left())
    return CoupleAppreciation(
        **row, from_name="Du", is_own=True,
    )


@router.post("/wertschaetzung/gesehen")
async def mark_appreciations_seen(
    couple_id: UUID, current=Depends(get_current_user), pool=Depends(get_pool),
) -> dict[str, int]:
    """Hakt den „neu"-Hinweis ab. Wer sie geschrieben hat, erfaehrt davon nichts."""
    async with pool.acquire() as conn:
        anzahl = await cas_appr.mark_seen(conn, couple_id, current["user_id"])
    return {"seen": anzahl}


# ── Stimmungsbarometer ──────────────────────────────────────────────────────
# Zustand, kein Urteil: Wie geht es DIR gerade mit euch. Ohne Blindheitsregel - ein
# Barometer, das man erst nach eigener Eingabe sieht, waere keines.


@router.get("/barometer", response_model=CoupleBarometerState)
async def get_barometer(
    couple_id: UUID, current=Depends(get_current_user), pool=Depends(get_pool),
) -> CoupleBarometerState:
    async with pool.acquire() as conn:
        data = await cbar.load_state(conn, couple_id, current["user_id"])
    return CoupleBarometerState(**data)


@router.put("/barometer", response_model=CoupleBarometerState)
async def set_barometer(
    couple_id: UUID, body: CoupleBarometerSet,
    current=Depends(get_current_user), pool=Depends(get_pool),
) -> CoupleBarometerState:
    user_id = current["user_id"]
    async with pool.acquire() as conn:
        neu = await cbar.set_value(conn, couple_id, user_id, body.value, body.note)
        # Einmal am Tag zaehlt es - der Regler soll nicht zum Punkte-Automaten werden.
        await progress.award(conn, couple_id, user_id, "barometer_set",
                             neu["created_at"].date().isoformat())

        # Nur bei einem deutlichen Absacken. Eine Meldung bei jedem Regler-Zupfen waere
        # Laerm; ein Einbruch ist genau der Moment, in dem Nachfragen etwas bringt.
        vorher = neu.get("previous_value")
        if vorher is not None and vorher - neu["value"] >= cbar.DROP_NOTICE:
            await notify.to_partner(conn, couple_id, user_id, notify.barometer_dropped())

        data = await cbar.load_state(conn, couple_id, user_id)
    return CoupleBarometerState(**data)
